index out of range

```
❯ k logs -n accuknox-agents   discovery-engine-5469f46499-cbtbv --previous
1:47PM INF usr/src/knox/src/libs/common.go:76 > BUILD-INFO: commit:ce80984, branch: dev, date: 2023-03-29T03:25:50Z, version: 0.1
1:47PM INF usr/src/knox/src/main.go:33 > NETWORK-POLICY: {OperationMode:1 OperationTrigger:5 CronJobTimeInterval:@every 0h0m10s OneTimeJobTimeSelection: NetworkLogLimit:10000 NetworkLogFrom:kubearmor NetworkLogFile: NetworkPolicyTo:db NetworkPolicyDir:./ NsFilter:[] NsNotFilter:[kube-system] NetPolicyTypes:3 NetPolicyRuleTypes:1023 NetPolicyCIDRBits:32 NetLogFilters:[] NetPolicyL3Level:1 NetPolicyL4Level:1 NetPolicyL7Level:1 NetSkipCertVerification:true}
1:47PM INF usr/src/knox/src/main.go:34 > CILIUM: {HubbleURL:hubble-relay.kube-system.svc.cluster.local HubblePort:80}
1:47PM INF usr/src/knox/src/main.go:35 > SYSTEM-POLICY: {OperationMode:1 OperationTrigger:5 CronJobTimeInterval:@every 0h0m10s OneTimeJobTimeSelection: SystemLogLimit:10000 SystemLogFrom:kubearmor SystemLogFile: SystemPolicyTo:db SystemPolicyDir:./ SysPolicyTypes:7 DeprecateOldMode:true SystemLogFilters:[] NsFilter:[] NsNotFilter:[kube-system] FromSourceFilter:[knoxAutoPolicy] ProcessFromSource:true FileFromSource:true}
1:47PM INF usr/src/knox/src/main.go:36 > KUBEARMOR: {KubeArmorRelayURL:kubearmor.kube-system.svc.cluster.local KubeArmorRelayPort:32767}
1:47PM INF usr/src/knox/src/networkpolicy/networkPolicy.go:2362 > Auto network policy discovery cron job started
1:47PM INF usr/src/knox/src/systempolicy/systemPolicy.go:1492 > Auto system policy discovery cron job started
1:47PM INF usr/src/knox/src/observability/observability.go:79 > Observability cron job started
1:47PM INF usr/src/knox/src/observability/observability.go:97 > Publisher cron job started
1:47PM INF usr/src/knox/src/recommendpolicy/recommendPolicy.go:73 > Recommended policy cron job started
panic: runtime error: index out of range [0] with length 0

goroutine 38 [running]:
github.com/accuknox/auto-policy-discovery/src/cluster.GetKubearmorRelayURL()
        /usr/src/knox/src/cluster/k8sClientHandler.go:497 +0x185
github.com/accuknox/auto-policy-discovery/src/systempolicy.StartSystemLogRcvr()
        /usr/src/knox/src/systempolicy/systemPolicy.go:1466 +0x53
created by github.com/accuknox/auto-policy-discovery/src/systempolicy.StartSystemCronJob
        /usr/src/knox/src/systempolicy/systemPolicy.go:1481 +0x25
```

--- a/src/cluster/k8sClientHandler.go
+++ b/src/cluster/k8sClientHandler.go
@@ -491,7 +491,8 @@ func GetKubearmorRelayURL() string {
 		log.Error().Msg(err.Error())
 		return ""
 	}
-	if pods == nil {
+	if pods == nil || len(pods.Items) == 0 {
+		log.Error().Msgf("Unable to find kubearmor-relay")
 		return ""
 	}
 	namespace = pods.Items[0].Namespace

--- a/src/plugin/kubearmor.go
+++ b/src/plugin/kubearmor.go
@@ -19,6 +19,7 @@ import (
 	"github.com/accuknox/auto-policy-discovery/src/types"
 	pb "github.com/kubearmor/KubeArmor/protobuf"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 )
 
 // Global Variable
@@ -313,13 +314,18 @@ func ConvertKubeArmorLogToKnoxSystemLog(relayLog *pb.Alert) (types.KnoxSystemLog
 func ConnectKubeArmorRelay(cfg types.ConfigKubeArmorRelay) *grpc.ClientConn {
 	addr := net.JoinHostPort(cfg.KubeArmorRelayURL, cfg.KubeArmorRelayPort)
 
-	conn, err := grpc.Dial(addr, grpc.WithInsecure())
+	// Check for kubearmor-relay with 30s timeout
+	ctx, cf1 := context.WithTimeout(context.Background(), time.Second*30)
+	defer cf1()
+
+	// Blocking grpc Dial: in case of a bad connection, fails with timeout
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
 	if err != nil {
-		log.Error().Msg("err connecting kubearmor relay. " + err.Error())
+		log.Error().Msg("Error connecting kubearmor relay: " + err.Error())
 		return nil
 	}
 
-	log.Info().Msg("connected to kubearmor relay " + addr)
+	log.Info().Msg("Connected to kubearmor relay " + addr)
 	return conn
 }
 
@@ -327,7 +333,7 @@ func GetSystemAlertsFromKubeArmorRelay(trigger int) []*pb.Alert {
 	results := []*pb.Alert{}
 	KubeArmorRelayLogsMutex.Lock()
 	if len(KubeArmorRelayLogs) == 0 {
-		log.Info().Msgf("KubeArmor Relay traffic flow not exist")
+		log.Info().Msgf("KubeArmor Relay traffic flow does not exist")
 		KubeArmorRelayLogsMutex.Unlock()
 		return results
 	}

--- a/src/systempolicy/systemPolicy.go
+++ b/src/systempolicy/systemPolicy.go
@@ -1464,6 +1464,17 @@ func StartSystemLogRcvr() {
 	for {
 		if cfg.GetCfgSystemLogFrom() == "kubearmor" {
 			url := cluster.GetKubearmorRelayURL()
+			if url == "" {
+				log.Error().Msg("kubearmor-relay url not found, retrying...")
+				for i := 0; i < types.Maxtries; i++ {
+					time.Sleep(10 * time.Second)
+					url = cluster.GetKubearmorRelayURL()
+					if url != "" {
+						break
+					}
+				}
+				return
+			}
 			plugin.StartKubeArmorRelay(SystemStopChan, types.ConfigKubeArmorRelay{
 				KubeArmorRelayURL:  url,
 				KubeArmorRelayPort: cfg.CurrentCfg.ConfigKubeArmorRelay.KubeArmorRelayPort,

--- a/src/types/constants.go
+++ b/src/types/constants.go
@@ -44,4 +44,7 @@ const (
 	// K8sNetworkPolicy
 	K8sNwPolicyAPIVersion = "networking.k8s.io/v1"
 	K8sNwPolicyKind       = "NetworkPolicy"
+
+	// max no. of tries to connect to kubearmor-relay
+	Maxtries = 6
 )

